In [567]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

# Input data files are available in the read-only "../input/" directory
# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory

import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

# You can write up to 20GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using "Save & Run All" 
# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session

/kaggle/input/anime-dataset-2022/Anime.csv


- **Anime Recommender System**
- **Content Based Filtering**
- **Text Vectorization**
- **Bag of Words**
- **Cosine_similarity**

In [568]:
import warnings
warnings.filterwarnings('ignore')

# Import data
data_m = pd.read_csv("../input/anime-dataset-2022/Anime.csv")
pd.set_option("max_columns", 200)

In [569]:
data_m.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18495 entries, 0 to 18494
Data columns (total 17 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Rank             18495 non-null  int64  
 1   Name             18495 non-null  object 
 2   Japanese_name    7938 non-null   object 
 3   Type             18495 non-null  object 
 4   Episodes         9501 non-null   float64
 5   Studio           12018 non-null  object 
 6   Release_season   4116 non-null   object 
 7   Tags             18095 non-null  object 
 8   Rating           15364 non-null  float64
 9   Release_year     18112 non-null  float64
 10  End_year         2854 non-null   float64
 11  Description      18491 non-null  object 
 12  Content_Warning  1840 non-null   object 
 13  Related_Mange    7627 non-null   object 
 14  Related_anime    10063 non-null  object 
 15  Voice_actors     15309 non-null  object 
 16  staff            13005 non-null  object 
dtypes: float64(4

In [570]:
data_m.head()

Rank                                               Name  \
0     1  Demon Slayer: Kimetsu no Yaiba - Entertainment...   
1     2                     Fruits Basket the Final Season   
2     3                                    Mo Dao Zu Shi 3   
3     4                   Fullmetal Alchemist: Brotherhood   
4     5                Attack on Titan 3rd Season: Part II   

                                     Japanese_name   Type  Episodes  \
0                    Kimetsu no Yaiba: Yuukaku-hen  TV          NaN   
1                          Fruits Basket the Final  TV         13.0   
2                       The Founder of Diabolism 3  Web        12.0   
3   Hagane no Renkinjutsushi: Full Metal Alchemist  TV         64.0   
4             Shingeki no Kyojin Season 3: Part II  TV         10.0   

              Studio Release_season  \
0           ufotable          Fall    
1  TMS Entertainment         Spring   
2   B.C MAY PICTURES            NaN   
3              Bones         Spring   
4         WIT Studio         Spring   

                                                Tags  Rating  Release_year  \
0  Action, Adventure, Fantasy, Shounen, Demons, H...    4.60        2021.0   
1  Drama, Fantasy, Romance, Shoujo, Animal Transf...    4.60        2021.0   
2  Fantasy, Ancient China, Chinese Animation, Cul...    4.58        2021.0   
3  Action, Adventure, Drama, Fantasy, Mystery, Sh...    4.58        2009.0   
4  Action, Fantasy, Horror, Shounen, Dark Fantasy...    4.57        2019.0   

   End_year                                        Description  \
0       NaN  'Tanjiro and his friends accompany the Hashira...   
1       NaN                  'The final arc of Fruits Basket.'   
2       NaN               'The third season of Mo Dao Zu Shi.'   
3    2010.0  "The foundation of alchemy is based on the law...   
4       NaN  'The battle to retake Wall Maria begins now! W...   

                                     Content_Warning  \
0                                  Explicit Violence   
1  Emotional Abuse,, Mature Themes,, Physical Abu...   
2                                                NaN   
3  Animal Abuse,, Mature Themes,, Violence,, Dome...   
4                    Cannibalism,, Explicit Violence   

                                       Related_Mange  \
0                     Demon Slayer: Kimetsu no Yaiba   
1               Fruits Basket, Fruits Basket Another   
2  Grandmaster of Demonic Cultivation: Mo Dao Zu ...   
3  Fullmetal Alchemist, Fullmetal Alchemist (Ligh...   
4  Attack on Titan, Attack on Titan: End of the W...   

                                       Related_anime  \
0  Demon Slayer: Kimetsu no Yaiba, Demon Slayer: ...   
1  Fruits Basket 1st Season, Fruits Basket 2nd Se...   
2                   Mo Dao Zu Shi 2, Mo Dao Zu Shi Q   
3  Fullmetal Alchemist: Brotherhood Specials, Ful...   
4  Attack on Titan, Attack on Titan 2nd Season, A...   

                                        Voice_actors  \
0  Inosuke Hashibira : Yoshitsugu Matsuoka, Nezuk...   
1  Akito Sohma : Maaya Sakamoto, Kyo Sohma : Yuum...   
2  Lan Wangji, Wei Wuxian, Jiang Cheng, Jin Guang...   
3  Alphonse Elric : Rie Kugimiya, Edward Elric : ...   
4  Armin Arlelt : Marina Inoue, Eren Jaeger : Yuu...   

                                               staff  
0  Koyoharu Gotouge : Original Creator, Haruo Sot...  
1  Natsuki Takaya : Original Creator, Yoshihide I...  
2  Mo Xiang Tong Xiu : Original Creator, Xiong Ke...  
3  Hiromu Arakawa : Original Creator, Yasuhiro Ir...  
4  Hajime Isayama : Original Creator, Tetsurou Ar...

In [571]:
data_m.tail()

Rank                                               Name Japanese_name  \
18490  18491  Qin Shi Mingyue: Canghai Hengliu Xiaomeng Spec...           NaN   
18491  18492                      Yi Tang Juchang: Sanguo Yanyi           NaN   
18492  18493          Fenghuang Ji Xiang Yu Qingming Shanghe Tu           NaN   
18493  18494                      Chengshi Jiyi Wo Men de Jieri           NaN   
18494  18495  Heisei Inu Monogatari Bow: Genshi Inu Monogata...           NaN   

        Type  Episodes                        Studio Release_season  \
18490  Web         2.0  Sparkly Key Animation Studio            NaN   
18491  TV        108.0                           NaN            NaN   
18492  TV         13.0                           NaN            NaN   
18493  TV          NaN                           NaN            NaN   
18494  Movie       NaN              Nippon Animation            NaN   

                                                    Tags  Rating  \
18490  Action, Ancient China, Chinese Animation, Hist...     NaN   
18491                                  Chinese Animation     NaN   
18492  Chinese Animation, Family Friendly, Short Epis...     NaN   
18493  Chinese Animation, Family Friendly, Short Epis...     NaN   
18494                        Comedy, Slice of Life, Dogs     NaN   

       Release_year  End_year  \
18490        2020.0       NaN   
18491        2010.0       NaN   
18492        2020.0       NaN   
18493        2020.0       NaN   
18494        1994.0       NaN   

                                             Description Content_Warning  \
18490  Special episodes of Qin Shi Mingyue: Canghai H...             NaN   
18491                 No synopsis yet - check back soon!             NaN   
18492                 No synopsis yet - check back soon!             NaN   
18493                 No synopsis yet - check back soon!             NaN   
18494                 No synopsis yet - check back soon!             NaN   

      Related_Mange                     Related_anime Voice_actors staff  
18490           NaN  Qin Shi Mingyue: Canghai Hengliu          NaN   NaN  
18491           NaN                               NaN          NaN   NaN  
18492           NaN                               NaN          NaN   NaN  
18493           NaN                               NaN          NaN   NaN  
18494           NaN         Heisei Inu Monogatari Bow          NaN   NaN

Looking at the observations

In [572]:
data_m["Tags"][2]

'Fantasy, Ancient China, Chinese Animation, Cultivation, Xianxia, Based on a Web Novel'

In [573]:
data_m["Description"][2]

"'The third season of Mo Dao Zu Shi.'"

In [574]:
data_m["Content_Warning"][2]

nan

In [575]:
data_m["Related_Mange"][2]

'Grandmaster of Demonic Cultivation: Mo Dao Zu Shi (Novel), The Master of Diabolism'

In [576]:
data_m["staff"][0]

'Koyoharu Gotouge : Original Creator, Haruo Sotozaki : Director, Akira Matsushima : Character Design, Aimer : Song Performance'

# Filtering the columns for Model 

In [577]:
#### The features required
# Data for ml algo
# Rank
# Name
# Type
# Tags
# Description
# Staff "Get Original Creator"

In [578]:
data_m = data_m[["Rank", "Name", "Type", "Tags", "Description", "Content_Warning", "staff"]]

In [579]:
# No duplicates
data_m.duplicated().sum()

0

In [580]:
# Checking if any NaN exist
data_m.isnull().sum()

Rank                   0
Name                   0
Type                   0
Tags                 400
Description            4
Content_Warning    16655
staff               5490
dtype: int64

In [581]:
data_m["Type"].value_counts()

TV       5446
Movie    3577
Web      2488
OVA      2235
Music    2165
Other     990
DVD S     911
TV Sp     683
Name: Type, dtype: int64

# Data Cleaning

In [582]:
# Type variable has unnecessary space
data_m[data_m["Type"] == "TV   "]

Rank                                               Name   Type  \
0          1  Demon Slayer: Kimetsu no Yaiba - Entertainment...  TV      
1          2                     Fruits Basket the Final Season  TV      
3          4                   Fullmetal Alchemist: Brotherhood  TV      
4          5                Attack on Titan 3rd Season: Part II  TV      
5          6                                     Jujutsu Kaisen  TV      
...      ...                                                ...    ...   
18481  18482                                    Goblin Slayer 2  TV      
18488  18489        Shachiku-san wa Youjo Yuurei ni Iyasaretai.  TV      
18491  18492                      Yi Tang Juchang: Sanguo Yanyi  TV      
18492  18493          Fenghuang Ji Xiang Yu Qingming Shanghe Tu  TV      
18493  18494                      Chengshi Jiyi Wo Men de Jieri  TV      

                                                    Tags  \
0      Action, Adventure, Fantasy, Shounen, Demons, H...   
1      Drama, Fantasy, Romance, Shoujo, Animal Transf...   
3      Action, Adventure, Drama, Fantasy, Mystery, Sh...   
4      Action, Fantasy, Horror, Shounen, Dark Fantasy...   
5      Action, Horror, Shounen, Curse, Exorcists, Mon...   
...                                                  ...   
18481  Action, Adventure, Fantasy, Seinen, Dark Fanta...   
18488  Comedy, Slice of Life, Ghosts, Iyashikei, Non-...   
18491                                  Chinese Animation   
18492  Chinese Animation, Family Friendly, Short Epis...   
18493  Chinese Animation, Family Friendly, Short Epis...   

                                             Description  \
0      'Tanjiro and his friends accompany the Hashira...   
1                      'The final arc of Fruits Basket.'   
3      "The foundation of alchemy is based on the law...   
4      'The battle to retake Wall Maria begins now! W...   
5      'Although Yuji Itadori looks like your average...   
...                                                  ...   
18481  It's fall, and the village's harvest festival ...   
18488  The story follows the daily life of Fushihara,...   
18491                 No synopsis yet - check back soon!   
18492                 No synopsis yet - check back soon!   
18493                 No synopsis yet - check back soon!   

                                         Content_Warning  \
0                                      Explicit Violence   
1      Emotional Abuse,, Mature Themes,, Physical Abu...   
3      Animal Abuse,, Mature Themes,, Violence,, Dome...   
4                        Cannibalism,, Explicit Violence   
5                                      Explicit Violence   
...                                                  ...   
18481                                                NaN   
18488                                                NaN   
18491                                                NaN   
18492                                                NaN   
18493                                                NaN   

                                                   staff  
0      Koyoharu Gotouge : Original Creator, Haruo Sot...  
1      Natsuki Takaya : Original Creator, Yoshihide I...  
3      Hiromu Arakawa : Original Creator, Yasuhiro Ir...  
4      Hajime Isayama : Original Creator, Tetsurou Ar...  
5      Gege Akutami : Original Creator, Seong-Hu Park...  
...                                                  ...  
18481                      Kumo Kagyu : Original Creator  
18488                     Imari Arita : Original Creator  
18491                                                NaN  
18492                                                NaN  
18493                                                NaN  

[5446 rows x 7 columns]

In [583]:
# Filtering the type column by removing spaces and putting it back into the dataframe
temp = data_m["Type"].apply(lambda x: [" ".join(x.split()) for i in x])

In [584]:
def type_(input):
    L = []
    for i in input:
        L.append(i[0])
    return L

In [585]:
data_m["Type"]= pd.Series(type_(temp))

In [586]:
# unnecessary space are removed
data_m[data_m["Type"] == "TV"]

Rank                                               Name Type  \
0          1  Demon Slayer: Kimetsu no Yaiba - Entertainment...   TV   
1          2                     Fruits Basket the Final Season   TV   
3          4                   Fullmetal Alchemist: Brotherhood   TV   
4          5                Attack on Titan 3rd Season: Part II   TV   
5          6                                     Jujutsu Kaisen   TV   
...      ...                                                ...  ...   
18481  18482                                    Goblin Slayer 2   TV   
18488  18489        Shachiku-san wa Youjo Yuurei ni Iyasaretai.   TV   
18491  18492                      Yi Tang Juchang: Sanguo Yanyi   TV   
18492  18493          Fenghuang Ji Xiang Yu Qingming Shanghe Tu   TV   
18493  18494                      Chengshi Jiyi Wo Men de Jieri   TV   

                                                    Tags  \
0      Action, Adventure, Fantasy, Shounen, Demons, H...   
1      Drama, Fantasy, Romance, Shoujo, Animal Transf...   
3      Action, Adventure, Drama, Fantasy, Mystery, Sh...   
4      Action, Fantasy, Horror, Shounen, Dark Fantasy...   
5      Action, Horror, Shounen, Curse, Exorcists, Mon...   
...                                                  ...   
18481  Action, Adventure, Fantasy, Seinen, Dark Fanta...   
18488  Comedy, Slice of Life, Ghosts, Iyashikei, Non-...   
18491                                  Chinese Animation   
18492  Chinese Animation, Family Friendly, Short Epis...   
18493  Chinese Animation, Family Friendly, Short Epis...   

                                             Description  \
0      'Tanjiro and his friends accompany the Hashira...   
1                      'The final arc of Fruits Basket.'   
3      "The foundation of alchemy is based on the law...   
4      'The battle to retake Wall Maria begins now! W...   
5      'Although Yuji Itadori looks like your average...   
...                                                  ...   
18481  It's fall, and the village's harvest festival ...   
18488  The story follows the daily life of Fushihara,...   
18491                 No synopsis yet - check back soon!   
18492                 No synopsis yet - check back soon!   
18493                 No synopsis yet - check back soon!   

                                         Content_Warning  \
0                                      Explicit Violence   
1      Emotional Abuse,, Mature Themes,, Physical Abu...   
3      Animal Abuse,, Mature Themes,, Violence,, Dome...   
4                        Cannibalism,, Explicit Violence   
5                                      Explicit Violence   
...                                                  ...   
18481                                                NaN   
18488                                                NaN   
18491                                                NaN   
18492                                                NaN   
18493                                                NaN   

                                                   staff  
0      Koyoharu Gotouge : Original Creator, Haruo Sot...  
1      Natsuki Takaya : Original Creator, Yoshihide I...  
3      Hiromu Arakawa : Original Creator, Yasuhiro Ir...  
4      Hajime Isayama : Original Creator, Tetsurou Ar...  
5      Gege Akutami : Original Creator, Seong-Hu Park...  
...                                                  ...  
18481                      Kumo Kagyu : Original Creator  
18488                     Imari Arita : Original Creator  
18491                                                NaN  
18492                                                NaN  
18493                                                NaN  

[5446 rows x 7 columns]

In [587]:
data_m['Content_Warning'].isnull()

0        False
1        False
2         True
3        False
4        False
         ...  
18490     True
18491     True
18492     True
18493     True
18494     True
Name: Content_Warning, Length: 18495, dtype: bool

In [588]:
data_m['Content_Warning'].fillna('', inplace = True)

In [589]:
data_m['Content_Warning']

0                                        Explicit Violence
1        Emotional Abuse,, Mature Themes,, Physical Abu...
2                                                         
3        Animal Abuse,, Mature Themes,, Violence,, Dome...
4                          Cannibalism,, Explicit Violence
                               ...                        
18490                                                     
18491                                                     
18492                                                     
18493                                                     
18494                                                     
Name: Content_Warning, Length: 18495, dtype: object

In [590]:
data_m['Content_Warning'].isnull()

0        False
1        False
2        False
3        False
4        False
         ...  
18490    False
18491    False
18492    False
18493    False
18494    False
Name: Content_Warning, Length: 18495, dtype: bool

In [591]:
data_m["Tags"] = data_m["Tags"] + ', ' + data_m['Content_Warning']

In [592]:
data_m["Name"]

0        Demon Slayer: Kimetsu no Yaiba - Entertainment...
1                           Fruits Basket the Final Season
2                                          Mo Dao Zu Shi 3
3                         Fullmetal Alchemist: Brotherhood
4                      Attack on Titan 3rd Season: Part II
                               ...                        
18490    Qin Shi Mingyue: Canghai Hengliu Xiaomeng Spec...
18491                        Yi Tang Juchang: Sanguo Yanyi
18492            Fenghuang Ji Xiang Yu Qingming Shanghe Tu
18493                        Chengshi Jiyi Wo Men de Jieri
18494    Heisei Inu Monogatari Bow: Genshi Inu Monogata...
Name: Name, Length: 18495, dtype: object

In [593]:
# data_m.drop('Content_Warning', axis=1, inplace=True)

In [594]:
# Basically observations with no tags and useless Description
# We will drop these from data_m
filtering_1 = data_m[ (data_m["Description"] == "'No synopsis yet - check back soon!'") | (data_m["Tags"].isnull())].index
data_m = data_m.drop(filtering_1)

In [595]:
data_m.isnull().sum()

Rank                  0
Name                  0
Type                  0
Tags                  0
Description           4
Content_Warning       0
staff              3794
dtype: int64

In [596]:
# Let us drop the observations which are of type Music, OVA , TV Sp. Since they are mostly a filler anime
filtering_2 = data_m[ (data_m["Type"] == "Music") | (data_m["Type"] == "OVA") | (data_m["Type"] == "TV Sp")].index
data_m.drop(filtering_2,inplace = True)

In [597]:
data_m.isnull().sum()

Rank                  0
Name                  0
Type                  0
Tags                  0
Description           4
Content_Warning       0
staff              3112
dtype: int64

In [598]:
filtering_3 = data_m[ (data_m["Description"] == "No synopsis yet - check back soon!") & (data_m["Tags"].isnull())].index
data_m.drop(filtering_3,inplace=True)

In [599]:
data_m[data_m["Tags"].isnull()]

Empty DataFrame
Columns: [Rank, Name, Type, Tags, Description, Content_Warning, staff]
Index: []

In [600]:
# tags do play a crucial role in recommendations. So the observations with Rank 7911, 8325, 8342, 10186.
# WILL BE REMOVED
filtering_4 = data_m[data_m["Tags"].isnull()].index
data_m.drop(filtering_4,inplace=True)

In [601]:
data_m.isnull().sum()

Rank                  0
Name                  0
Type                  0
Tags                  0
Description           4
Content_Warning       0
staff              3112
dtype: int64

In [602]:
data_m.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 10969 entries, 0 to 18494
Data columns (total 7 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   Rank             10969 non-null  int64 
 1   Name             10969 non-null  object
 2   Type             10969 non-null  object
 3   Tags             10969 non-null  object
 4   Description      10965 non-null  object
 5   Content_Warning  10969 non-null  object
 6   staff            7857 non-null   object
dtypes: int64(1), object(6)
memory usage: 943.6+ KB


In [603]:
# Upon looking at the Description column. We have noticed that not all the observations have descriptions
# If we look at it most of them are 'No synopsis yet - check back soon!'   OR
# No synopsis yet - check back soon!    as the description.

data_m.drop("Description", axis=1, inplace=True)

In [604]:
data_m.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 10969 entries, 0 to 18494
Data columns (total 6 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   Rank             10969 non-null  int64 
 1   Name             10969 non-null  object
 2   Type             10969 non-null  object
 3   Tags             10969 non-null  object
 4   Content_Warning  10969 non-null  object
 5   staff            7857 non-null   object
dtypes: int64(1), object(5)
memory usage: 857.9+ KB


Upon going thorough the staff variable.
- We have noticed that there are 5490 missing values. We could get rid of the column entirely.
- However in the world of Anime. Original Creaters like Eiichiro Oda (One Piece), Hajime Isayama (Attack on Titan), Masashi Kishimoto (Naruto), many more have a great Manga sales throughout the world.

What is so great about having sales for considering Original Creaters name?

There are people who are willing to watch the anime based on the anime Original Creater. Hence, we will use it as one of the features

**Therefore, we will only keeping the observations with staff variable present which containd the Original Creater Name.**

In [605]:
data_m["staff"].isna().sum()

3112

# Filtering of Staff variable

In [606]:
# Filtering out the values that has staff
data_m_mini = data_m[data_m["staff"].notnull()]

In [607]:
data_m_mini.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 7857 entries, 0 to 18489
Data columns (total 6 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   Rank             7857 non-null   int64 
 1   Name             7857 non-null   object
 2   Type             7857 non-null   object
 3   Tags             7857 non-null   object
 4   Content_Warning  7857 non-null   object
 5   staff            7857 non-null   object
dtypes: int64(1), object(5)
memory usage: 429.7+ KB


In [608]:
data_m_mini["staff"][0]

'Koyoharu Gotouge : Original Creator, Haruo Sotozaki : Director, Akira Matsushima : Character Design, Aimer : Song Performance'

In [609]:
# We need to convert these to dict
data_m_mini["staff"][1]

'Natsuki Takaya : Original Creator, Yoshihide Ibata : Director & Episode Director & Storyboard, Taku Kishimoto : Screenplay & Series Composition, Masaru Yokoyama : Music, Masaru Shindou : Character Design & Chief Animation Director, Baek-Ryun Chae : Photography Director, Youko Koyama : Art Director, Mika Sugawara : Color Design'

In [610]:
data_m_mini

Rank                                               Name   Type  \
0          1  Demon Slayer: Kimetsu no Yaiba - Entertainment...     TV   
1          2                     Fruits Basket the Final Season     TV   
2          3                                    Mo Dao Zu Shi 3    Web   
3          4                   Fullmetal Alchemist: Brotherhood     TV   
4          5                Attack on Titan 3rd Season: Part II     TV   
...      ...                                                ...    ...   
18484  18485                                           Tooriame  Movie   
18486  18487                                   Better back then  Movie   
18487  18488                                        Make My Day    Web   
18488  18489        Shachiku-san wa Youjo Yuurei ni Iyasaretai.     TV   
18489  18490                                          Nostalgia  Movie   

                                                    Tags  \
0      Action, Adventure, Fantasy, Shounen, Demons, H...   
1      Drama, Fantasy, Romance, Shoujo, Animal Transf...   
2      Fantasy, Ancient China, Chinese Animation, Cul...   
3      Action, Adventure, Drama, Fantasy, Mystery, Sh...   
4      Action, Fantasy, Horror, Shounen, Dark Fantasy...   
...                                                  ...   
18484                                 Abstract, Shorts,    
18486     Shorts, Stop Motion Animation, Original Work,    
18487                  Horror, Monsters, Original Work,    
18488  Comedy, Slice of Life, Ghosts, Iyashikei, Non-...   
18489  Drama, No Dialogue, Shorts, Stop Motion Animat...   

                                         Content_Warning  \
0                                      Explicit Violence   
1      Emotional Abuse,, Mature Themes,, Physical Abu...   
2                                                          
3      Animal Abuse,, Mature Themes,, Violence,, Dome...   
4                        Cannibalism,, Explicit Violence   
...                                                  ...   
18484                                                      
18486                                                      
18487                                                      
18488                                                      
18489                                                      

                                                   staff  
0      Koyoharu Gotouge : Original Creator, Haruo Sot...  
1      Natsuki Takaya : Original Creator, Yoshihide I...  
2      Mo Xiang Tong Xiu : Original Creator, Xiong Ke...  
3      Hiromu Arakawa : Original Creator, Yasuhiro Ir...  
4      Hajime Isayama : Original Creator, Tetsurou Ar...  
...                                                  ...  
18484  Tomoko Takayama : Director, Juri Ito : Music, ...  
18486  Naoaki Shibuta : Director, Chikara Uemizutaru ...  
18487  Yasuo Ootagaki : Original Creator, Makoto Kane...  
18488                     Imari Arita : Original Creator  
18489                         Tomoyasu Murata : Director  

[7857 rows x 6 columns]

In [611]:
# Let use reset the index for the staff conversion from a Series str to a Series of dict
data_m_mini.reset_index(inplace = True)

In [612]:
#### SAMPLE #####
# Convert to dict
try_1 = data_m_mini["staff"][1].split(",")
try_1 = {k.split(":")[0].strip() : k.split(":")[1].strip() for k in try_1}
print(try_1)

{'Natsuki Takaya': 'Original Creator', 'Yoshihide Ibata': 'Director & Episode Director & Storyboard', 'Taku Kishimoto': 'Screenplay & Series Composition', 'Masaru Yokoyama': 'Music', 'Masaru Shindou': 'Character Design & Chief Animation Director', 'Baek-Ryun Chae': 'Photography Director', 'Youko Koyama': 'Art Director', 'Mika Sugawara': 'Color Design'}


In [613]:
# a Series of  str to Dict
for i in range (0,len(data_m_mini["staff"])):
    try_1 = data_m_mini["staff"][i].split(",")
    data_m_mini["staff"][i] = {k.split(":")[0].strip() : k.split(":")[1].strip() for k in try_1}

In [614]:
data_m_mini["staff"][1]

{'Natsuki Takaya': 'Original Creator',
 'Yoshihide Ibata': 'Director & Episode Director & Storyboard',
 'Taku Kishimoto': 'Screenplay & Series Composition',
 'Masaru Yokoyama': 'Music',
 'Masaru Shindou': 'Character Design & Chief Animation Director',
 'Baek-Ryun Chae': 'Photography Director',
 'Youko Koyama': 'Art Director',
 'Mika Sugawara': 'Color Design'}

In [615]:
data_m_mini["staff"][1].keys()

dict_keys(['Natsuki Takaya', 'Yoshihide Ibata', 'Taku Kishimoto', 'Masaru Yokoyama', 'Masaru Shindou', 'Baek-Ryun Chae', 'Youko Koyama', 'Mika Sugawara'])

In [616]:
data_m_mini["staff"][1].values()

dict_values(['Original Creator', 'Director & Episode Director & Storyboard', 'Screenplay & Series Composition', 'Music', 'Character Design & Chief Animation Director', 'Photography Director', 'Art Director', 'Color Design'])

In [617]:
# Converting values to keys and keys to values in Dict
for i in range (0,len(data_m_mini["staff"])):
    dict = data_m_mini["staff"][i]
    data_m_mini["staff"][i] = {value:key for key, value in dict.items()}

In [618]:
data_m_mini.iloc[8000:8023,3:4]

Empty DataFrame
Columns: [Type]
Index: []

In [619]:
# Filtering for Original Creators in staff Column
for i in range (0,len(data_m_mini["staff"])):
    data_m_mini["staff"][i] = { key:value for (key,value) in data_m_mini["staff"][i].items() if key  == "Original Creator"}

In [620]:
data_m_mini.iloc[1:40,3:4]

Type
1      TV
2     Web
3      TV
4      TV
5      TV
6      TV
7      TV
8   Movie
9      TV
10  Movie
11     TV
12     TV
13     TV
14  Movie
15  Movie
16     TV
17     TV
18     TV
19     TV
20    Web
21  Movie
22    Web
23     TV
24     TV
25     TV
26     TV
27  Movie
28     TV
29    Web
30     TV
31     TV
32     TV
33     TV
34    Web
35     TV
36     TV
37    Web
38  Movie
39     TV

In [621]:
# filtering for dict with values
filtering_5 = data_m_mini[data_m_mini["staff"] == {}].index
data_m_mini.drop(filtering_5,inplace = True)

In [622]:
# Reset index aagin
data_m_mini.reset_index(inplace = True)

In [623]:
# Keeping the values of Original Creator
for i in range (0,len(data_m_mini["staff"])):
        data_m_mini["staff"][i] = (data_m_mini["staff"][i])["Original Creator"]

*We have succesfully extracted the Original Creater Name.* 

In [624]:
data_m_mini

level_0  index   Rank  \
0           0      0      1   
1           1      1      2   
2           2      2      3   
3           3      3      4   
4           4      4      5   
...       ...    ...    ...   
2809     7838  18443  18444   
2810     7850  18481  18482   
2811     7851  18483  18484   
2812     7854  18487  18488   
2813     7855  18488  18489   

                                                   Name   Type  \
0     Demon Slayer: Kimetsu no Yaiba - Entertainment...     TV   
1                        Fruits Basket the Final Season     TV   
2                                       Mo Dao Zu Shi 3    Web   
3                      Fullmetal Alchemist: Brotherhood     TV   
4                   Attack on Titan 3rd Season: Part II     TV   
...                                                 ...    ...   
2809                                      The Imaginary  Movie   
2810                                    Goblin Slayer 2     TV   
2811                        Peleliu: Rakuen no Guernica  Other   
2812                                        Make My Day    Web   
2813        Shachiku-san wa Youjo Yuurei ni Iyasaretai.     TV   

                                                   Tags  \
0     Action, Adventure, Fantasy, Shounen, Demons, H...   
1     Drama, Fantasy, Romance, Shoujo, Animal Transf...   
2     Fantasy, Ancient China, Chinese Animation, Cul...   
3     Action, Adventure, Drama, Fantasy, Mystery, Sh...   
4     Action, Fantasy, Horror, Shounen, Dark Fantasy...   
...                                                 ...   
2809                                 Based on a Novel,    
2810  Action, Adventure, Fantasy, Seinen, Dark Fanta...   
2811  Action, Shounen, Island, War, World War 2, Bas...   
2812                  Horror, Monsters, Original Work,    
2813  Comedy, Slice of Life, Ghosts, Iyashikei, Non-...   

                                        Content_Warning              staff  
0                                     Explicit Violence   Koyoharu Gotouge  
1     Emotional Abuse,, Mature Themes,, Physical Abu...     Natsuki Takaya  
2                                                        Mo Xiang Tong Xiu  
3     Animal Abuse,, Mature Themes,, Violence,, Dome...     Hiromu Arakawa  
4                       Cannibalism,, Explicit Violence     Hajime Isayama  
...                                                 ...                ...  
2809                                                         A. F. Harrold  
2810                                                            Kumo Kagyu  
2811                                                       Masao Hiratsuka  
2812                                                        Yasuo Ootagaki  
2813                                                           Imari Arita  

[2814 rows x 8 columns]

In [625]:
data_m_mini["Tags"] = data_m_mini["Tags"] +", "+ data_m_mini["staff"]

In [626]:
data_m_mini.drop(labels = "staff",axis = 1, inplace= True)

In [627]:
data_m.drop("staff",axis = 1, inplace = True)

In [628]:
data_m["Tags"][0]

'Action, Adventure, Fantasy, Shounen, Demons, Historical, Martial Arts, Orphans, Siblings, Swordplay, Based on a Manga, Explicit Violence, Explicit Violence'

In [629]:
data_m_mini["Tags"][0]

'Action, Adventure, Fantasy, Shounen, Demons, Historical, Martial Arts, Orphans, Siblings, Swordplay, Based on a Manga, Explicit Violence, Explicit Violence, Koyoharu Gotouge'

In [630]:
data_m.set_index(['Rank','Name'], inplace=True)
data_m.update(data_m_mini.set_index(['Rank','Name']))
data_m.reset_index(inplace=True)
data_m

Rank                                               Name   Type  \
0          1  Demon Slayer: Kimetsu no Yaiba - Entertainment...     TV   
1          2                     Fruits Basket the Final Season     TV   
2          3                                    Mo Dao Zu Shi 3    Web   
3          4                   Fullmetal Alchemist: Brotherhood     TV   
4          5                Attack on Titan 3rd Season: Part II     TV   
...      ...                                                ...    ...   
10964  18491  Qin Shi Mingyue: Canghai Hengliu Xiaomeng Spec...    Web   
10965  18492                      Yi Tang Juchang: Sanguo Yanyi     TV   
10966  18493          Fenghuang Ji Xiang Yu Qingming Shanghe Tu     TV   
10967  18494                      Chengshi Jiyi Wo Men de Jieri     TV   
10968  18495  Heisei Inu Monogatari Bow: Genshi Inu Monogata...  Movie   

                                                    Tags  \
0      Action, Adventure, Fantasy, Shounen, Demons, H...   
1      Drama, Fantasy, Romance, Shoujo, Animal Transf...   
2      Fantasy, Ancient China, Chinese Animation, Cul...   
3      Action, Adventure, Drama, Fantasy, Mystery, Sh...   
4      Action, Fantasy, Horror, Shounen, Dark Fantasy...   
...                                                  ...   
10964  Action, Ancient China, Chinese Animation, Hist...   
10965                                Chinese Animation,    
10966  Chinese Animation, Family Friendly, Short Epis...   
10967  Chinese Animation, Family Friendly, Short Epis...   
10968                      Comedy, Slice of Life, Dogs,    

                                         Content_Warning  
0                                      Explicit Violence  
1      Emotional Abuse,, Mature Themes,, Physical Abu...  
2                                                         
3      Animal Abuse,, Mature Themes,, Violence,, Dome...  
4                        Cannibalism,, Explicit Violence  
...                                                  ...  
10964                                                     
10965                                                     
10966                                                     
10967                                                     
10968                                                     

[10969 rows x 5 columns]

In [631]:
for i in range(0,len(data_m["Tags"])):
    data_m["Tags"][i] = data_m["Tags"][i].replace(" ","")

In [632]:
for i in range(0,len(data_m["Tags"])):
    data_m["Tags"][i] = data_m["Tags"][i].replace(","," ")

In [633]:
data_m["Tags"] = data_m["Tags"].apply(lambda x: x.lower())

**Next let us make a recommendation model**

In [634]:
data = data_m.copy()

In [635]:
data

Rank                                               Name   Type  \
0          1  Demon Slayer: Kimetsu no Yaiba - Entertainment...     TV   
1          2                     Fruits Basket the Final Season     TV   
2          3                                    Mo Dao Zu Shi 3    Web   
3          4                   Fullmetal Alchemist: Brotherhood     TV   
4          5                Attack on Titan 3rd Season: Part II     TV   
...      ...                                                ...    ...   
10964  18491  Qin Shi Mingyue: Canghai Hengliu Xiaomeng Spec...    Web   
10965  18492                      Yi Tang Juchang: Sanguo Yanyi     TV   
10966  18493          Fenghuang Ji Xiang Yu Qingming Shanghe Tu     TV   
10967  18494                      Chengshi Jiyi Wo Men de Jieri     TV   
10968  18495  Heisei Inu Monogatari Bow: Genshi Inu Monogata...  Movie   

                                                    Tags  \
0      action adventure fantasy shounen demons histor...   
1      drama fantasy romance shoujo animaltransformat...   
2      fantasy ancientchina chineseanimation cultivat...   
3      action adventure drama fantasy mystery shounen...   
4      action fantasy horror shounen darkfantasy isol...   
...                                                  ...   
10964  action ancientchina chineseanimation historica...   
10965                                  chineseanimation    
10966     chineseanimation familyfriendly shortepisodes    
10967     chineseanimation familyfriendly shortepisodes    
10968                           comedy sliceoflife dogs    

                                         Content_Warning  
0                                      Explicit Violence  
1      Emotional Abuse,, Mature Themes,, Physical Abu...  
2                                                         
3      Animal Abuse,, Mature Themes,, Violence,, Dome...  
4                        Cannibalism,, Explicit Violence  
...                                                  ...  
10964                                                     
10965                                                     
10966                                                     
10967                                                     
10968                                                     

[10969 rows x 5 columns]

In [636]:
# data = data[data['Rank'] < 2000]

In [637]:
# data

# # Text Vectarization
# # Bag of words

In [638]:
from sklearn.feature_extraction.text import TfidfVectorizer
tfidf = TfidfVectorizer(max_features = 13500,stop_words="english")

In [639]:
tfidf.fit_transform(data["Tags"]).toarray()

array([[0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       ...,
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.]])

In [640]:
tfidf.fit_transform(data["Tags"]).toarray().shape

(10969, 1894)

In [641]:
vectors = tfidf.fit_transform(data["Tags"]).toarray()

In [642]:
tfidf.get_feature_names()[0:10]

['15thcentury',
 '16thcentury',
 '17thcentury',
 '18thcentury',
 '19thcentury',
 'abiumeda',
 'abstract',
 'acomomota',
 'acting',
 'action']

In [643]:
from sklearn.metrics.pairwise import cosine_similarity
similarity = cosine_similarity(vectors)
similarity

array([[1.        , 0.07308995, 0.02417408, ..., 0.        , 0.        ,
        0.        ],
       [0.07308995, 1.        , 0.01399412, ..., 0.        , 0.        ,
        0.        ],
       [0.02417408, 0.01399412, 1.        , ..., 0.11438871, 0.11438871,
        0.        ],
       ...,
       [0.        , 0.        , 0.11438871, ..., 1.        , 1.        ,
        0.        ],
       [0.        , 0.        , 0.11438871, ..., 1.        , 1.        ,
        0.        ],
       [0.        , 0.        , 0.        , ..., 0.        , 0.        ,
        1.        ]])

In [644]:
similarity.shape

(10969, 10969)

In [645]:
def Recommender(user_input, rec_type):
    Index_of_anime = data[data["Name"] == user_input].index[0]
    Similarity_score = similarity[Index_of_anime]
    Sorted_scores = sorted(list(enumerate(Similarity_score)),reverse = True,key= lambda x: x[1])[1:]
    Recommended_Anime = []
    
    for i in Sorted_scores:
        if data.iloc[i[0]].Rank > 2000:
            continue
        if len(Recommended_Anime) > 5:
            break
        if rec_type is None:
            Recommended_Anime.append(data.iloc[i[0]].Name)
        elif data.iloc[i[0]].Type == rec_type:
            Recommended_Anime.append(data.iloc[i[0]].Name)
    return Recommended_Anime

In [646]:
Recommender("Mo Dao Zu Shi 3", None)

['Douluo Dalu 2',
 'Mo Dao Zu Shi 2',
 'Mo Dao Zu Shi',
 'Ling Jian Zun 4th Season',
 'Ling Jian Zun 5th Season',
 'Ling Jian Zun 3rd Season']

In [647]:
Recommender("Yaoshenji 2nd Season", None)

['Yaoshenji 4th Season',
 'Yaoshenji 2nd Season',
 'Yaoshenji 5th Season',
 'Star Martial God Technique',
 'Yaoshenji',
 'Douluo Dalu 2']

In [648]:
Recommender("Gunslinger Stratos: The Animation", None)

['Time Jam: Valerian & Laureline',
 'City Hunter Movie: Shinjuku Private Eyes',
 'Tales of Zestiria: The X (2017)',
 'Tales of Zestiria: The X',
 'Persona 3 the Movie 4: Winter of Rebirth',
 'Persona 3 the Movie 1: Spring of Birth']

In [649]:
Recommender("Fighting Foodons", None)

['Digimon Adventure Movie 2: Bokura no War Game',
 'Pokemon XY&Z',
 'Fox Spirit Matchmaker Season 3',
 'Yakitate!! Japan',
 'Pokemon Journeys',
 'Toriko TV']

In [650]:
Recommender("Jujutsu Kaisen", None)

['Jujutsu Kaisen 0',
 'Kill Bill Chapter 3: The Origin of O-Ren',
 'Dorohedoro',
 'Baki: Son of Ogre',
 'A Certain Magical Index',
 'Shiki Specials']

In [651]:
Recommender("Jujutsu Kaisen", "Movie")

['Jujutsu Kaisen 0',
 'Kill Bill Chapter 3: The Origin of O-Ren',
 'Kizumonogatari Part 2: Nekketsu',
 'Made in Abyss Movie: Dawn of the Deep Soul',
 'Thunderbolt Fantasy Movie: The Sword of Life and Death',
 'Kizumonogatari Part 1: Tekketsu']

In [652]:
Recommender("Demon Slayer: Kimetsu no Yaiba - Entertainment District Arc", "TV")

['Demon Slayer: Kimetsu no Yaiba',
 'The Legend of the Legendary Heroes',
 'A Certain Magical Index',
 'Ga-Rei-Zero',
 'Demon Slayer: Kimetsu no Yaiba - Mugen Train',
 'Claymore']

In [653]:
Recommender("Naruto Shippuden", "TV")

['Naruto',
 'Ashita no Joe 2',
 'Dragon Ball Z',
 'Dragon Ball Z Kai: The Final Chapters',
 'Hajime no Ippo: Rising',
 'Hajime no Ippo: New Challenger']

In [654]:
Recommender("Fullmetal Alchemist: Brotherhood", "TV")

['Fullmetal Alchemist',
 "JoJo's Bizarre Adventure: Diamond is Unbreakable",
 'ERASED',
 'The Heroic Legend of Arslan',
 'The Seven Deadly Sins',
 'Tiger & Bunny']

#                             Thank you